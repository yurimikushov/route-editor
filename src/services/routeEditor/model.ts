type RouteEditorState = {
  route: Array<Address>
}

type RouteEditor = {
  route: Array<Address>
  addAddress: (newAddress: NewAddress) => void
  changeAddress: (id: string, newPoint: Point) => void
  deleteAddress: (id: string) => void
  updateRoute: (route: Array<Address>) => void
}

type NewAddress = Omit<Address, 'id'>

type Address = {
  id: string
  name: string
  description: string
  point: Point
}

type Point = {
  lat: number
  lon: number
}

export type { RouteEditorState, RouteEditor, NewAddress, Address, Point }
