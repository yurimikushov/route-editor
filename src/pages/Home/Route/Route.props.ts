type RouteProps = {
  className?: string
}

export default RouteProps
