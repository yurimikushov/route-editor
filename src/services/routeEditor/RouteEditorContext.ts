import isNull from 'lodash/isNull'
import { createContext, useContext } from 'react'
import { RouteEditor } from './model'

const RouteEditorContext = createContext<RouteEditor | null>(null)

const useRouteEditor = () => {
  const routeEditor = useContext(RouteEditorContext)

  if (isNull(routeEditor)) {
    throw new Error(
      '[Route editor] Route editor should be created before it usage'
    )
  }

  return routeEditor
}

export default RouteEditorContext
export { useRouteEditor }
