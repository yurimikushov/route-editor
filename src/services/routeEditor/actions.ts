import {
  ADD_ADDRESS,
  CHANGE_ADDRESS_BEGIN,
  CHANGE_ADDRESS_SUCCESS,
  CHANGE_ADDRESS_FAIL,
  DELETE_ADDRESS,
  UPDATE_ROUTE,
} from './constants'
import { Address, NewAddress, Point } from './model'

const addAddress = (newAddress: NewAddress) =>
  ({
    type: ADD_ADDRESS,
    payload: {
      newAddress,
    },
  } as const)

const changeAddressBegin = (id: string, newPoint: Point) =>
  ({
    type: CHANGE_ADDRESS_BEGIN,
    payload: {
      id,
      newPoint,
    },
  } as const)

const changeAddressSuccess = (id: string, newAddress: NewAddress) =>
  ({
    type: CHANGE_ADDRESS_SUCCESS,
    payload: {
      id,
      newAddress,
    },
  } as const)

const changeAddressFail = (id: string) =>
  ({
    type: CHANGE_ADDRESS_FAIL,
    payload: {
      id,
    },
  } as const)

const deleteAddress = (id: string) =>
  ({
    type: DELETE_ADDRESS,
    payload: {
      id,
    },
  } as const)

const updateRoute = (route: Array<Address>) =>
  ({
    type: UPDATE_ROUTE,
    payload: {
      route,
    },
  } as const)

type AddAddress = ReturnType<typeof addAddress>
type ChangeAddressBegin = ReturnType<typeof changeAddressBegin>
type ChangeAddressSuccess = ReturnType<typeof changeAddressSuccess>
type ChangeAddressFail = ReturnType<typeof changeAddressFail>
type DeleteAddress = ReturnType<typeof deleteAddress>
type UpdateRoute = ReturnType<typeof updateRoute>

type Action =
  | AddAddress
  | ChangeAddressBegin
  | ChangeAddressSuccess
  | ChangeAddressFail
  | DeleteAddress
  | UpdateRoute

export {
  addAddress,
  changeAddressBegin,
  changeAddressSuccess,
  changeAddressFail,
  deleteAddress,
  updateRoute,
}
export type { Action }
