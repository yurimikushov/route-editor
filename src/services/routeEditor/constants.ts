export const ADD_ADDRESS = 'routeEditor/ADD_ADDRESS'
export const CHANGE_ADDRESS_BEGIN = 'routeEditor/CHANGE_ADDRESS_BEGIN'
export const CHANGE_ADDRESS_SUCCESS = 'routeEditor/CHANGE_ADDRESS_SUCCESS'
export const CHANGE_ADDRESS_FAIL = 'routeEditor/CHANGE_ADDRESS_FAIL'
export const DELETE_ADDRESS = 'routeEditor/DELETE_ADDRESS'
export const UPDATE_ROUTE = 'routeEditor/UPDATE_ROUTE'
