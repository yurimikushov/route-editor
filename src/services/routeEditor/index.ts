export { default as RouteEditorProvider } from './RouteEditorProvider'
export { useRouteEditor } from './RouteEditorContext'
export type { RouteEditor, Address, Point } from './model'
