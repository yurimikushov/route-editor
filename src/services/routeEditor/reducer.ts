import { nanoid } from 'nanoid'
import map from 'lodash/map'
import reject from 'lodash/reject'
import { Action } from './actions'
import {
  ADD_ADDRESS,
  CHANGE_ADDRESS_BEGIN,
  CHANGE_ADDRESS_SUCCESS,
  CHANGE_ADDRESS_FAIL,
  DELETE_ADDRESS,
  UPDATE_ROUTE,
} from './constants'
import { RouteEditorState } from './model'

const routeEditorReducer = (
  state: RouteEditorState,
  action: Action
): RouteEditorState => {
  switch (action.type) {
    case ADD_ADDRESS: {
      const { newAddress } = action.payload

      return {
        route: [
          ...state.route,
          {
            ...newAddress,
            id: nanoid(),
          },
        ],
      }
    }
    case CHANGE_ADDRESS_BEGIN: {
      const { id, newPoint } = action.payload

      return {
        route: map(state.route, (address) => {
          if (address.id === id) {
            return {
              id: id,
              name: 'Адрес обновляется...',
              description: '',
              point: newPoint,
            }
          }

          return address
        }),
      }
    }
    case CHANGE_ADDRESS_SUCCESS: {
      const { id, newAddress } = action.payload

      return {
        route: map(state.route, (address) => {
          if (address.id === id) {
            return {
              ...newAddress,
              id,
            }
          }

          return address
        }),
      }
    }
    case CHANGE_ADDRESS_FAIL: {
      const { id } = action.payload

      return {
        route: map(state.route, (address) => {
          if (address.id === id) {
            return {
              ...address,
              name: 'Не удалось обновить адрес',
              description: '',
            }
          }

          return address
        }),
      }
    }
    case DELETE_ADDRESS: {
      const { id } = action.payload

      return {
        route: reject(state.route, { id }),
      }
    }
    case UPDATE_ROUTE: {
      const { route } = action.payload

      return {
        route,
      }
    }
    default:
      return state
  }
}

export default routeEditorReducer
