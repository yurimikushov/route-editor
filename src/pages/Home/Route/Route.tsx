import { FC } from 'react'
import cn from 'classnames'
import isEmpty from 'lodash/isEmpty'
import { useRouteEditor } from 'services/routeEditor'
import DraggableList from 'components/DraggableList'
import Point from './Point'
import Placeholder from './Placeholder'
import RouteProps from './Route.props'

const Route: FC<RouteProps> = ({ className }) => {
  const { route, deleteAddress, updateRoute } = useRouteEditor()

  return (
    <>
      {!isEmpty(route) ? (
        <DraggableList
          className={cn(className, 'overflow-y-scroll pl-1 pr-1 -mr-2')}
          draggableClassName='-mx-1 rounded-lg border-4 border-dashed border-gray-200'
          list={route}
          renderItem={(address, isDraggable) => (
            <Point
              key={address.id}
              point={address}
              isDraggable={isDraggable}
              onDelete={() => deleteAddress(address.id)}
            />
          )}
          onUpdate={updateRoute}
        />
      ) : (
        <Placeholder />
      )}
    </>
  )
}

export default Route
