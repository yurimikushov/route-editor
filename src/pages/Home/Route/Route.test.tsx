import renderer from 'react-test-renderer'
import noop from 'lodash/noop'
import { useRouteEditor } from 'services/routeEditor'
import Route from './Route'

// eslint-disable-next-line react/display-name
jest.mock('./Point/Point', () => (props: { className?: string }) => (
  <div {...props}></div>
))

jest.mock('services/routeEditor')

const runMockUseRouteEditor = (result: ReturnType<typeof useRouteEditor>) => {
  // @ts-expect-error bad typing
  useRouteEditor.mockImplementation(() => result)
}

it(`should render points when route isn't empty`, () => {
  runMockUseRouteEditor({
    route: [
      {
        name: 'Balalayka',
        description: 'Some description',
        point: {
          lat: 0,
          lon: 0,
        },
      },
      {
        name: 'Vodka',
        description: 'Some description',
        point: {
          lat: 1,
          lon: 1,
        },
      },
    ],
    addPoint: noop,
    changePoint: noop,
    deletePoint: noop,
    updateRoute: noop,
  })

  const tree = renderer.create(<Route />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should render placeholder when route is empty', () => {
  runMockUseRouteEditor({
    route: [],
    addPoint: noop,
    changePoint: noop,
    deletePoint: noop,
    updateRoute: noop,
  })

  const tree = renderer.create(<Route />).toJSON()
  expect(tree).toMatchSnapshot()
})
