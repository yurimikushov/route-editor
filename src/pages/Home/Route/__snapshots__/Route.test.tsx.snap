// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render placeholder when route is empty 1`] = `
<div>
  Placeholder
</div>
`;

exports[`should render points when route isn't empty 1`] = `
<ul
  className="py-2"
  data-rbd-droppable-context-id="0"
  data-rbd-droppable-id="draggable-list"
>
  <li
    aria-describedby="rbd-hidden-text-0-hidden-text-0"
    data-rbd-drag-handle-context-id="0"
    data-rbd-drag-handle-draggable-id="1"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="1"
    draggable={false}
    onDragStart={[Function]}
    onTransitionEnd={null}
    role="button"
    style={
      Object {
        "transform": null,
        "transition": null,
      }
    }
    tabIndex={-1}
  >
    <div
      onDelete={[Function]}
      point={
        Object {
          "description": "Some description",
          "id": "1",
          "name": "Balalayka",
          "point": Object {
            "lat": 0,
            "lon": 0,
          },
        }
      }
    />
  </li>
  <li
    aria-describedby="rbd-hidden-text-0-hidden-text-0"
    data-rbd-drag-handle-context-id="0"
    data-rbd-drag-handle-draggable-id="2"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="2"
    draggable={false}
    onDragStart={[Function]}
    onTransitionEnd={null}
    role="button"
    style={
      Object {
        "transform": null,
        "transition": null,
      }
    }
    tabIndex={-1}
  >
    <div
      onDelete={[Function]}
      point={
        Object {
          "description": "Some description",
          "id": "2",
          "name": "Vodka",
          "point": Object {
            "lat": 1,
            "lon": 1,
          },
        }
      }
    />
  </li>
</ul>
`;
